addons/product/models/product_catalog_mixin.py

```diff
@@ -0,0 +1,93 @@
+
+
+
+class ProductCatalogMixin(models.AbstractModel):
+    """ This mixin should be inherited when the model should be able to work
+    with the product catalog.
+    """
+    _name = 'product.catalog.mixin'
+    _description = 'Product Catalog Mixin'
+
+    def action_add_from_catalog(self):
+        kanban_view_id = self.env.ref('product.product_view_kanban_catalog').id
+        search_view_id = self.env.ref('product.product_view_search_catalog').id
+        additional_context = self._get_action_add_from_catalog_extra_context()
+        return {
+            'type': 'ir.actions.act_window',
+            'name': _('Products'),
+            'res_model': 'product.product',
+            'views': [(kanban_view_id, 'kanban'), (False, 'form')],
+            'search_view_id': [search_view_id, 'search'],
+            'domain': self._get_product_catalog_domain(),
+            'context': {**self.env.context, **additional_context},
+        }
+
+    def _default_order_line_values(self):
+        return {
+            'quantity': 0,
+            'readOnly': self._is_readonly() if self else False,
+        }
+
+    def _get_product_catalog_domain(self):
+        """Get the domain to search for products in the catalog.
+
+        For a model that uses products that has to be hidden in the catalog, it must override this
+        method and extend the appropriate domain.
+        :returns: A list of tuples that represents a domain.
+        :rtype: list
+        """
+        return [('company_id', 'in', [self.company_id.id, False])]
+
+    def _get_product_catalog_record_lines(self, product_ids):
+        """ Returns the record's lines grouped by product.
+        Must be overrided by each model using this mixin.
+
+        :param list product_ids: The ids of the products currently displayed in the product catalog.
+        :rtype: dict
+        """
+        return {}
+
+    def _get_product_catalog_order_line_info(self, product_ids, **kwargs):
+        """ Returns products information to be shown in the catalog.
+        :param list product_ids: The products currently displayed in the product catalog, as a list
+                                 of `product.product` ids.
+        :rtype: dict
+        :return: A dict with the following structure:
+            {
+                'productId': int
+                'quantity': float (optional)
+                'price': float
+                'readOnly': bool (optional)
+            }
+        """
+        order_line_info = {}
+        default_data = self._default_order_line_values()
+
+        for product, order_lines in self._get_product_catalog_record_lines(product_ids).items():
+            order_line_info[product.id] = order_lines._get_product_catalog_lines_data(**kwargs)
+            product_ids.remove(product.id)
+
+        products = self.env['product.product'].browse(product_ids) - self.order_line.product_id
+        product_data = self._get_product_catalog_order_data(products, **kwargs)
```

Comment:
`_get_product_catalog_order_data` should also exist in this mixin with a docstring on how it's mandatory to override